Export All Playlists Feature

mvsolves opened the pull request:
This feature simply allows a user to 'Export All Playlists' when right clicking on the 'Playlist' tab within the library.

Some things to note:
- For each exported playlist, a window pops up asking what format to export the playlist in
- Pressing the 'cancel' button will cancel all other playlist exports.

mvsolves committed:
Create 'Export All Playlists' menu option

Add necessary methods and definitions to create the 'Export All Playlists'
method. 'Export All Playlists' still works like 'Export Playlist' method.
Add textField to display this option to user.

mvsolves committed:
Feat: Create 'slotExportPlaylists'

Create a slot that iterates over each playlist and exports them. Each
iteration essentially does what 'slotExportPlaylist' is meant to do.

mvsolves committed:
Add 'Export All Playlists' in 'Playlists' menu

Moved feature 'Export All Playlists' to the 'Playlist' menu when right clicking. Before, you were able to 'Export All Playlists' when right clicking on a random playlist, which does not make sense.

mvsolves committed:
Declare variables outside 'slotExportPlaylists'

Moved declaration of variables outside the for loop within the
'slotExportPlaylists' method.

daschuer commented:
Thank you for your first pull request. 

I think this suits to be a last minute PR for the 2.4-beta. I have changed the target branch accordingly. 
Please merge the 2.4 branch and resolve the conflict.

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -600,6 +600,101 @@ void BasePlaylistFeature::slotExportPlaylist() {
+void BasePlaylistFeature::slotExportPlaylists() {
+    const PlaylistDAO::HiddenType hidden = PlaylistDAO::PLHT_NOT_HIDDEN;
+
+    QList<QPair<int, QString>> allPlaylists = m_playlistDao.getPlaylists(hidden);
+
+    QPair<int, QString> thePair;
```
Such an anonymous QPair is always a pain in reviews. I would be happy if you would go the extra mile to create a named struct for it.

mvsolves replied:
Can you explain a bit on what you mean by an 'anonymous QPair'?

mvsolves replied:
> Such an anonymous QPair is always a pain in reviews. I would be happy if you would go the extra mile to create a named struct for it.

Still confused on this

daschuer replied:
A QPair can contain everything. 
It's members "first' and "second' are badly named, in every context. 

The proposed code lints that by hiding this ugly detail of QPair. 

The alternative to QPair is a custom struct, with properly named members. This should be used whenever the Pair escapes the scope of just being a return value.

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -600,6 +600,101 @@ void BasePlaylistFeature::slotExportPlaylist() {
+
+    QList<QPair<int, QString>> allPlaylists = m_playlistDao.getPlaylists(hidden);
+
+    QPair<int, QString> thePair;
+    int playlistId;
+    QString playlistName;
+
+    for (int i = 0; i < allPlaylists.length(); i++) {
+        thePair = allPlaylists.at(i);
+        playlistId = thePair.first;
+        playlistName = m_playlistDao.getPlaylistName(playlistId);
```
Please use here a range based loop, because i is only used to access the allPlaylists list. Here a syntax with the pair, that removes some of the pain. 
```suggestion
    for (const auto& [playlistId, playlistName] : allPlaylists) {
```

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -600,6 +600,101 @@ void BasePlaylistFeature::slotExportPlaylist() {
+        }
+
+        // replace separator character with something generic
+        playlistName = playlistName.replace(QDir::separator(), kUnsafeFilenameReplacement);
```
Please consider which part of this body can be extracted into a common function called from slotExportPlaylists() and slotExportPlaylist()

mvsolves replied:
Completed. Let me know what u think

daschuer commented on src/mixxxmainwindow.cpp:
```diff
@@ -745,16 +747,21 @@ void MixxxMainWindow::connectMenuBar() {
-                m_pCoreServices->getTrackCollectionManager().get(),
-                &TrackCollectionManager::startLibraryScan,
-                Qt::UniqueConnection);
+        // MV: This internally disables the rescan library
```
Is this a debug leftover?

ronso0 replied:
@mvsolves  the entire first commit is unrelated and can be removed.
When you rebase, please try to find more compact commit messages (title). If a commit needs more explanation use the message body which can also have longer lines.
IIRC we linked some resources on the wiki, but I don't find it hehe, but this looks good, too https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/

mvsolves replied:
@daschuer yep. It was from a previous branch I was working on (still learning about branches) and somehow created a branch that was on top of the other branch.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -129,14 +129,20 @@ void WMainMenuBar::initialize() {
     pLibraryRescan->setCheckable(false);
+
+    // MV: This externally (visually) disables the pLibraryRescan icon
+    // pLibraryRescan->setDisabled(true);
```
???

daschuer reviewed (changes requested):
Thank you for this good start. I have left some comments.

daschuer commented:
Before merge, we need your permission to distribute your changes with Mixxx.
Please sign: https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
And comment here when done.

ronso0 commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -611,84 +602,94 @@ void BasePlaylistFeature::slotExportPlaylist() {
+    for (int i = 0; i < allPlaylists.length(); i++) {
+        QPair thePair = allPlaylists.at(i);
+        int playlistId = thePair.first;
+        QString playlistName = m_playlistDao.getPlaylistName(playlistId);
```
This additional lookup _per playlist_ is redundant, you already get the name as `thePair.second`.
some more meaningful names would are always welcome, `thePair` > `idName`?

mvsolves replied:
Solved by changing that loop to range based loop: 
```cpp
    for (const auto& [playlistId, playlistName] : allPlaylists) {
        ...
    }

```

ronso0 commented on src/library/trackset/playlistfeature.cpp:
```diff
@@ -91,7 +92,7 @@ void PlaylistFeature::onRightClickChild(
     menu.addAction(m_pImportPlaylistAction);
     menu.addAction(m_pExportPlaylistAction);
-    menu.addAction(m_pExportPlaylistsAction);
+    // menu.addAction(m_pExportPlaylistsAction);
```
please remove unused code

mvsolves reviewed (commented):
Here is what I fixed so far

mvsolves commented:
> Before merge, we need your permission to distribute your changes with Mixxx.
> Please sign: https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
> And comment here when done.

Just completed

daschuer commented on CMakeLists.txt:
```diff
@@ -1,4 +1,9 @@
 cmake_minimum_required(VERSION 3.16)
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON) # does not produce the json file
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON CACHE INTERNAL "") # works
+
+set(CMAKE_OSX_ARCHITECTURES "x86_64")
```
This does not belong into the CMakeLists.txt you may create a toolchain file with this info or pass it on the command line.

mmvdepaul replied:
Can you explain how to create this toolchain file? I'm very new to CMake

mvsolves replied:
Can you explain how to create this toolchain file? I'm very new to CMake

daschuer commented:
> Made fixes based off pull request comments

Is not a suitable commit message. It is a good practice to use an imperative that refects the code change and keep the commit that small that the message still covers all changes. 
In your case of the changes above: 
* Remove stray blank lines

ronso0 commented on src/library/trackset/playlistfeature.cpp:
```diff
@@ -91,6 +92,7 @@ void PlaylistFeature::onRightClickChild(
     menu.addSeparator();
     menu.addAction(m_pImportPlaylistAction);
     menu.addAction(m_pExportPlaylistAction);
+    // menu.addAction(m_pExportPlaylistsAction);
```
Please don't resolve conversations yourself, it's up to the reviewer to confirm follow-up changes.

mvsolves replied:
Gotcha

ronso0 commented:
@mvsolves  Furthermore, please consider that every commit should build in case anyone has to do git bisect later on.
Please undo the cmakelist changes altogether (drop the first commit and 9aaaee0b382c82e5d80f8ec9d09e04529216a246, as well as the cmakelist changes in 33ec243edae772806c8c16a5625528a83cfd53c3), and try not to make no-op commits (like ed53e12099f80fd23cbc7b82510ce664f013f8ee which just duplicates existing `slotExportPlaylist`? simply combine it with the next commit or restructure your changes in a reasonable manner)
@daschuer do you agree on a rebase?

mvsolves commented:
> and try not to make no-op commits (like [ed53e12](https://github.com/mixxxdj/mixxx/commit/ed53e12099f80fd23cbc7b82510ce664f013f8ee) which just duplicates existing `slotExportPlaylist`? simply combine it with the next commit or restructure your changes in a reasonable manner)

What do you mean by 'no-op' commit?

mvsolves committed:
Create common function for 'slotExportPlaylists'

This new exportPlaylistsById() function is used for 'slotExportPlaylists' and 'slotExportPlaylist'

mvsolves commented on src/library/trackset/baseplaylistfeature.h:
```diff
@@ -93,6 +93,7 @@ class BasePlaylistFeature : public BaseTrackSetFeature {
     // Get the QModelIndex of a playlist based on its id.  Returns QModelIndex()
     // on failure.
     QModelIndex indexFromPlaylistId(int playlistId);
+    void exportPlaylistById(int playlistId);
```
Would you say this is a good spot to place function? Is this a clear function name?

mvsolves reviewed (commented):
Added a common function that is used in both 'slotExportPlaylist' and 'slotExportPlaylists' function

mvsolves committed:
Change to range based loop when iterating playlist

Remove multiple lined declarations to a single for loop line. Also keep
PlaylistDAO::PLHT_NOT_HIDDEN within the same line as the 'allPlaylists'
variable declaration.

mvsolves committed:
Remove 'Export All Playlists' when right clicking specific playlist

mvsolves commented:
So I spent some time with rebase and finally moved commits, split commits, and renamed commits. Kind of messy but now I know how to do them.

But I have one more issue before working on the two requested changes (which I feel would be better to discuss on the [Export multiple playlists](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help/topic/Export.20Multiple.20Playlists.20Feature/near/326705169) zulip chat.

ronso0 commented:
Thank you, the comments make sense now.

Though, the first four, unrelated commits are still present and cause conflicts with the base branch. Please drop them during rebase, which also allows you to drop the last one92a547170b3dce580504fb97cd4869c148762197

mvsolves commented:
> Thank you, the comments make sense now.
> 
> Though, the first four, unrelated commits are still present and cause conflicts with the base branch. Please drop them during rebase, which also allows you to drop the last one92a547170b3dce580504fb97cd4869c148762197

This is what I did. Then I did ```git push -f``` and now it says its missing

ronso0 commented:
great! now there's a typo, see CI logs, for example https://github.com/mixxxdj/mixxx/actions/runs/4130258542/jobs/7136783261
